libs/utils/energy_model.py

```diff
@@ -774,7 +774,102 @@ def _find_core_groups(cls, target):
+
+        This uses the energy_model sysctl added by EAS patches to exposes
+        the frequency domains, together with a tuple of capacity, frequency
+        and active power for each CPU. This feature is not upstream in mainline
+        Linux (as of v4.17), and only exists in Android kernels later than
+        android-4.14.
+
+        Wrt. idle states - the EnergyModel constructed won't be aware of
+        any power data or topological dependencies for entering "cluster"
+        idle states since the simplified model has no such concept.
+
+        Initialises only Active States for CPUs and clears all other levels.
+
+        :param target: Devlib target object to read filesystem from. Must have
+                       cpufreq and cpuidle modules enabled.
+        :returns: Constructed EnergyModel object based on the parameters
+                  reported by the target.
+        """
+        if 'cpuidle' not in target.modules:
+            raise TargetError('Requires cpuidle devlib module. Please ensure '
+                               '"cpuidle" is listed in your target/test modules')
+
+        # Simplified EM on-disk format (for each frequency domain):
+        #    /sys/devices/system/cpu/energy_model/<frequency_domain>/..
+        #        ../capacity
+        #           contains a space-separated list of capacities in increasing order
+        #        ../cpus
+        #           cpulist-formatted representation of the cpus in the frequency domain
+        #        ../frequency
+        #           space-separated list of frequencies in corresponding order to capacities
+        #        ../power
+        #           space-separated list of power consumption in corresponding order to capacities
+        # taken together, the contents of capacity, frequency and power give you the required
+        # tuple for ActiveStates.
+        # hence, domain should be supplied as a glob, and fields should be
+        #   capacity, cpus, frequency, power
+
+        try:
+            sysfs_em = target.read_tree_values("/sys/devices/system/cpu/energy_model/", depth=3)
+        except:
+            raise TargetError('Simplified Energy Model not exposed in sysfs.')
+
+        cpu_to_fdom = {}
+        for fd, fields in sysfs_em.iteritems():
+            cpus = ranges_to_list(fields["cpus"])
+            for cpu in cpus:
+                cpu_to_fdom[cpu] = fd
+            sysfs_em[fd]['cpus'] = cpus
+            sysfs_em[fd]['frequency'] = map(int, sysfs_em[fd]['frequency'].split(' '))
+            sysfs_em[fd]['power'] =  map(int, sysfs_em[fd]['power'].split(' '))
+            sysfs_em[fd]['capacity'] = map(int, sysfs_em[fd]['capacity'].split(' '))
+
+        def read_active_states(cpu):
+            fd = sysfs_em[cpu_to_fdom[cpu]]
+            cstates = zip(fd['capacity'], fd['power'])
+            active_states = [ActiveState(c, p) for c, p in cstates]
+            return OrderedDict(zip(sorted(fd['frequency']), active_states))
```

Comment:
Shouldn't it be zip and then sorted ?
```
OrderedDict(sorted(zip(fd['frequency'], active_states)))
```
Unless it is assumed that fd['capacity'] is already sorted and fd['frequency'] is in random order.

Reply:
No, both should be sorted in the same order, so there is probably no need to call `sorted`

Reply:
Agreed, we should not sort - if the EM returns incorrect values, let's be able to see them.